tests/apollo/test_skvbc_reconfiguration.py

```diff
@@ -555,6 +555,228 @@ async def test_addRemove_command(self, bft_network):
+        op = operator.Operator(bft_network.config, client,  bft_network.builddir)
+        test_config = 'new_configuration_n_4_f_1_c_0'
+        await op.add_remove_with_wedge(test_config)
+        await self.verify_replicas_are_in_wedged_checkpoint(bft_network, checkpoint_before, range(bft_network.config.n))
+        await self.verify_last_executed_seq_num(bft_network, checkpoint_before)
+        await self.validate_stop_on_super_stable_checkpoint(bft_network, skvbc)
+        bft_network.stop_all_replicas()
+        # We now expect the replicas to start with a fresh new configuration which means that we
+        # need to see in the logs that isNewStorage() = true. Also,
+        # we expect to see that lastStableSeqNum = 0 (for example)
+        conf = TestConfig(n=4,
+                   f=1,
+                   c=0,
+                   num_clients=10,
+                   key_file_prefix=KEY_FILE_PREFIX,
+                   start_replica_cmd=start_replica_cmd,
+                   stop_replica_cmd=None,
+                   num_ro_replicas=0)
+        await bft_network.change_configuration(conf)
+        bft_network.start_all_replicas()
+        await self.validate_state_consistency(skvbc, key, val)
+        for r in bft_network.all_replicas():
+            last_stable_checkpoint = await bft_network.get_metric(r, bft_network, "Gauges", "lastStableSeqNum")
+            self.assertEqual(last_stable_checkpoint, 0)
+        await self.validate_state_consistency(skvbc, key, val)
+        for i in range(100):
+            await skvbc.write_known_kv()
+        for r in bft_network.all_replicas():
+            assert( r < 4 )
+            nb_fast_path = await bft_network.get_metric(r, bft_network, "Counters", "totalFastPaths")
+            self.assertGreater(nb_fast_path, 0)
+    
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 7)
+    async def test_remove_nodes_with_f_failures(self, bft_network):
+        """
+        In this test we show how a system operator can remove nodes (and thus reduce the cluster) from 7 nodes cluster
+        to 4 nodes cluster even when f nodes are not responding
+        For that the operator performs the following steps:
+        1. Stop 2 nodes (f=2)
+        2. Send a remove_node command - this command also wedges the system
+        3. Verify that all live (including the removed candidates) nodes have stopped
+        4. Load  a new configuration to the bft network
+        5. Rerun the cluster with only 4 nodes and make sure they succeed to perform transactions in fast path
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        client = bft_network.random_client()
+        for i in range(100):
+            await skvbc.write_known_kv()
+        # choose two replicas to crash and crash them
+        crashed_replicas = {5, 6} # For simplicity, we crash the last two replicas
```

Comment:
would it work also with random replicas? (I guess it will, but it will be a much stronger test if we choose the the crashed replicas randomly)

Reply:
It's bit complicated to test with random replicas failure instead of last two i.e. {5,6}. Let say, test chooses random nodes in {0,1,2,3} - for example 2,3. Then after i restart the cluster with (n=4,f=1), there are two nodes with empty state. So, they won't be able to catch up via state transfer. At least we should have 3 non-faulty nodes